thanks, gave a try to this branch with [Ex0bit](https://huggingface.co/Ex0bit/Qwen3.6-27B-PRISM-EAGLE3) GGUF generated as per EAGLE3 PR doc. Text generation crashes instantly like reported in https://github.com/ggml-org/llama.cpp/issues/24541

After patching `qwen35.cpp` as advised,  only single turn generation working but with very bad perf like mentioned before. Sorry no nice table from SpeedBench, but on my Tesla V100, Qwen3.6 27B performs (without MTP) at tg 27 tok/sec usually, but here I get from 15 to 17 tok/sec (promtp "generate pi estimator code in python").

if writing something as second turn, this crash occurs:

```
3.21.236.241 I slot launch_slot_: id  0 | task 47 | processing task, is_child = 0
3.21.236.272 I slot update_slots: id  0 | task 47 | Checking checkpoint with [10, 10] against 21...
3.21.289.034 W slot update_slots: id  0 | task 47 | restored context checkpoint (pos_min = 10, pos_max = 10, n_tokens = 11, n_past = 11, size = 149.665 MiB)
3.21.407.379 E init: the tokens of sequence 0 in the input batch have inconsistent sequence positions:
 - the last position stored in the memory module of the context (i.e. the KV cache) for sequence 0 is X = 9
 - the tokens for sequence 0 in the input batch have a starting position of Y = 11
 it is required that the sequence positions remain consecutive: Y = X + 1
3.21.407.388 E decode: failed to initialize batch
3.21.407.390 E llama_decode: failed to decode, ret = -1
3.21.407.392 E process: llama_decode(ctx_dft) failed rc=-1 (n_tokens=20, ubatch_pos[0]=11)
3.21.407.396 E srv  update_slots: failed to process speculative batch
3.21.504.498 E init: the tokens of sequence 0 in the input batch have inconsistent sequence positions:
 - the last position stored in the memory module of the context (i.e. the KV cache) for sequence 0 is X = 9
 - the tokens for sequence 0 in the input batch have a starting position of Y = 31
 it is required that the sequence positions remain consecutive: Y = X + 1
3.21.504.506 E decode: failed to initialize batch
3.21.504.507 E llama_decode: failed to decode, ret = -1
3.21.504.509 E process: llama_decode(ctx_dft) failed rc=-1 (n_tokens=21, ubatch_pos[0]=32)
3.21.504.512 E srv  update_slots: failed to process speculative batch
/home/user/llama.cpp-eagl3_qwen3.6/tools/server/server-context.cpp:3275: fatal error - please provide logs and repro in https://github.com/ggml-org/llama.cpp/pull/20277

3.21.565.303 E init: the tokens of sequence 0 in the input batch have inconsistent sequence positions:
```

--- a/common/common.cpp
+++ b/common/common.cpp
@@ -2034,7 +2034,7 @@ bool common_prompt_batch_decode(
 }
 
 size_t common_prompt_checkpoint::size() const {
-    return data_tgt.size() + data_dft.size();
+    return data_tgt.size() + data_dft.size() + data_dft_boundary_g_embd.size() * sizeof(float);
 }
 
 bool common_prompt_checkpoint::empty() const {
@@ -2049,6 +2049,7 @@ void common_prompt_checkpoint::clear() {
 
     data_tgt.clear();
     data_dft.clear();
+    data_dft_boundary_g_embd.clear();
 }
 
 void common_prompt_checkpoint::update_pos(
@@ -2138,4 +2139,5 @@ void common_prompt_checkpoint::clear_tgt() {
 
 void common_prompt_checkpoint::clear_dft() {
     data_dft.clear();
+    data_dft_boundary_g_embd.clear();
 }

--- a/common/common.h
+++ b/common/common.h
@@ -363,7 +363,7 @@ struct common_params_speculative {
 
     uint32_t need_n_rs_seq() const {
         bool needs_rs_seq = std::any_of(types.begin(), types.end(), [&](auto t) {
-            return t == COMMON_SPECULATIVE_TYPE_DRAFT_MTP;
+            return t == COMMON_SPECULATIVE_TYPE_DRAFT_MTP || t == COMMON_SPECULATIVE_TYPE_DRAFT_EAGLE3;
         });
 
         return needs_rs_seq ? draft.n_max : 0u;
@@ -1064,6 +1064,9 @@ struct common_prompt_checkpoint {
 
     std::vector<uint8_t> data_tgt;
     std::vector<uint8_t> data_dft;
+
+    // eagle3: deferred-boundary g_embd row stashed with the checkpoint
+    std::vector<float> data_dft_boundary_g_embd;
 
     size_t size() const;
 

--- a/common/speculative.cpp
+++ b/common/speculative.cpp
@@ -161,6 +161,10 @@ struct common_speculative_impl {
 
     virtual void accept(llama_seq_id seq_id, uint16_t n_accepted, bool is_other) = 0;
 
+    // eagle3: deferred-boundary g_embd stash for checkpoints (default: none)
+    virtual bool get_deferred_boundary(llama_seq_id /*seq_id*/, std::vector<float> & /*g_out*/) const { return false; }
+    virtual void set_deferred_boundary(llama_seq_id /*seq_id*/, llama_pos /*pos*/, const std::vector<float> & /*g*/) {}
+
     // true if this implementation requires the target context to extract post-norm embeddings
     virtual bool need_embd() const = 0;
 
@@ -839,6 +843,35 @@ struct common_speculative_impl_draft_eagle3 : public common_speculative_impl {
         std::memcpy(pending_g_last[seq_id].data(),
                     verify_g[seq_id].data() + (size_t) i_g * n_embd_dec,
                     (size_t) n_embd_dec * sizeof(float));
+    }
+
+    // we only need to stash the deferred boundary's g_embd row for recurrent/hybrid targets:
+    // their single-position checkpoints drop it on restore
+    bool need_boundary_stash() const {
+        const llama_model * model_tgt = llama_get_model(params.ctx_tgt);
+        return llama_model_is_recurrent(model_tgt) || llama_model_is_hybrid(model_tgt);
+    }
+
+    bool get_deferred_boundary(llama_seq_id seq_id, std::vector<float> & g_out) const override {
+        if (!need_boundary_stash()) {
+            return false;
+        }
+        if (seq_id < 0 || seq_id >= (llama_seq_id) n_seq || pending_pos_last[seq_id] < 0) {
+            return false;
+        }
+        g_out = pending_g_last[seq_id];
+        return true;
+    }
+
+    void set_deferred_boundary(llama_seq_id seq_id, llama_pos pos, const std::vector<float> & g) override {
+        if (!need_boundary_stash()) {
+            return;
+        }
+        if (seq_id < 0 || seq_id >= (llama_seq_id) n_seq || (int32_t) g.size() != n_embd_dec) {
+            return;
+        }
+        pending_pos_last[seq_id] = pos;
+        pending_g_last[seq_id]   = g;
     }
 
     bool need_embd() const override {
@@ -2115,6 +2148,30 @@ void common_speculative_accept(common_speculative * spec, llama_seq_id seq_id, u
         if (impl_other.get() != impl) {
             impl_other->accept(seq_id, n_accepted, true);
         }
+    }
+}
+
+bool common_speculative_get_deferred_boundary(common_speculative * spec, llama_seq_id seq_id, std::vector<float> & g_out) {
+    if (spec == nullptr) {
+        return false;
+    }
+
+    for (auto & impl : spec->impls) {
+        if (impl->get_deferred_boundary(seq_id, g_out)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void common_speculative_set_deferred_boundary(common_speculative * spec, llama_seq_id seq_id, llama_pos pos, const std::vector<float> & g) {
+    if (spec == nullptr) {
+        return;
+    }
+
+    for (auto & impl : spec->impls) {
+        impl->set_deferred_boundary(seq_id, pos, g);
     }
 }
 

--- a/common/speculative.h
+++ b/common/speculative.h
@@ -68,6 +68,10 @@ void common_speculative_draft(common_speculative * spec);
 // informs the speculative context that n_accepted tokens were accepted by the target model
 void common_speculative_accept(common_speculative * spec, llama_seq_id, uint16_t n_accepted);
 
+// eagle3: deferred-boundary g_embd stash for checkpoints (no-op for other draft types)
+bool common_speculative_get_deferred_boundary(common_speculative * spec, llama_seq_id seq_id, std::vector<float> & g_out);
+void common_speculative_set_deferred_boundary(common_speculative * spec, llama_seq_id seq_id, llama_pos boundary_pos, const std::vector<float> & g);
+
 // print statistics about the speculative decoding
 void common_speculative_print_stats(const common_speculative * spec);
 

--- a/common/speculative.h
+++ b/common/speculative.h
@@ -68,9 +68,9 @@ void common_speculative_draft(common_speculative * spec);
 // informs the speculative context that n_accepted tokens were accepted by the target model
 void common_speculative_accept(common_speculative * spec, llama_seq_id, uint16_t n_accepted);
 
-// eagle3: deferred-boundary g_embd stash for checkpoints (no-op for other draft types)
-bool common_speculative_get_deferred_boundary(common_speculative * spec, llama_seq_id seq_id, std::vector<float> & g_out);
-void common_speculative_set_deferred_boundary(common_speculative * spec, llama_seq_id seq_id, llama_pos boundary_pos, const std::vector<float> & g);
+// (optional) get/set internal state
+bool common_speculative_get_state(common_speculative * spec, llama_seq_id seq_id, std::vector<uint8_t> & data);
+void common_speculative_set_state(common_speculative * spec, llama_seq_id seq_id, const std::vector<uint8_t> & data);
 
 // print statistics about the speculative decoding
 void common_speculative_print_stats(const common_speculative * spec);

--- a/src/models/qwen35.cpp
+++ b/src/models/qwen35.cpp
@@ -156,6 +156,8 @@ llama_model_qwen35::graph::graph(const llama_model & model, const llm_graph_para
 
     // MTP/NextN layers are loaded as extra decoder blocks but not executed in the main pass.
     for (int il = 0; il < n_layer; ++il) {
+        res->t_layer_inp[il] = inpL;
+
         ggml_tensor * inpSA = inpL;
 
         cur = build_norm(inpL, model.layers[il].attn_norm, nullptr, LLM_NORM_RMS, il);

--- a/src/models/qwen35moe.cpp
+++ b/src/models/qwen35moe.cpp
@@ -179,6 +179,8 @@ llama_model_qwen35moe::graph::graph(const llama_model & model, const llm_graph_p
 
     // MTP/NextN layers are loaded as extra decoder blocks but not executed in the main pass.
     for (int il = 0; il < n_layer; ++il) {
+        res->t_layer_inp[il] = inpL;
+
         ggml_tensor * inpSA = inpL;
 
         cur = build_norm(inpL, model.layers[il].attn_norm, nullptr, LLM_NORM_RMS, il);

--- a/tools/server/server-context.cpp
+++ b/tools/server/server-context.cpp
@@ -2977,6 +2977,17 @@ private:
 
                                     bool do_reset = it == slot.prompt.checkpoints.rend();
 
+                                    // eagle3 draft is one position behind the target due to deferred boundary), so it
+                                    // can't resume from a checkpoint restored on a recurrent/hybrid target; re-process fully instead.
+                                    const bool spec_eagle3 = std::find(params_base.speculative.types.begin(), params_base.speculative.types.end(),
+                                                                       COMMON_SPECULATIVE_TYPE_DRAFT_EAGLE3) != params_base.speculative.types.end();
+                                    if (!do_reset && spec_eagle3 &&
+                                            (ctx_tgt_seq_rm_type == COMMON_CONTEXT_SEQ_RM_TYPE_FULL ||
+                                             ctx_tgt_seq_rm_type == COMMON_CONTEXT_SEQ_RM_TYPE_RS)) {
+                                        SLT_WRN(slot, "%s", "eagle3 draft cannot resume from a recurrent/hybrid checkpoint, forcing full re-processing\n");
+                                        do_reset = true;
+                                    }
+
                                     if (!do_reset) {
                                         // restore the context checkpoint
                                         it->load_tgt(ctx_tgt,       slot.id, LLAMA_STATE_SEQ_FLAGS_PARTIAL_ONLY);

--- a/tools/server/server-context.cpp
+++ b/tools/server/server-context.cpp
@@ -2157,6 +2157,8 @@ private:
 
         cur.update_tgt(ctx_tgt,       slot.id, LLAMA_STATE_SEQ_FLAGS_PARTIAL_ONLY);
         cur.update_dft(ctx_dft.get(), slot.id, LLAMA_STATE_SEQ_FLAGS_PARTIAL_ONLY);
+        // stash the draft's deferred boundary with the checkpoint (only eagle3 needs it; no-op otherwise)
+        common_speculative_get_deferred_boundary(spec.get(), slot.id, cur.data_dft_boundary_g_embd);
 
         SLT_INF(slot,
                 "created context checkpoint %d of %d (pos_min = %d, pos_max = %d, n_tokens = %" PRId64 ", size = %.3f MiB)\n",
@@ -2977,21 +2979,12 @@ private:
 
                                     bool do_reset = it == slot.prompt.checkpoints.rend();
 
-                                    // eagle3 draft is one position behind the target due to deferred boundary), so it
-                                    // can't resume from a checkpoint restored on a recurrent/hybrid target; re-process fully instead.
-                                    const bool spec_eagle3 = std::find(params_base.speculative.types.begin(), params_base.speculative.types.end(),
-                                                                       COMMON_SPECULATIVE_TYPE_DRAFT_EAGLE3) != params_base.speculative.types.end();
-                                    if (!do_reset && spec_eagle3 &&
-                                            (ctx_tgt_seq_rm_type == COMMON_CONTEXT_SEQ_RM_TYPE_FULL ||
-                                             ctx_tgt_seq_rm_type == COMMON_CONTEXT_SEQ_RM_TYPE_RS)) {
-                                        SLT_WRN(slot, "%s", "eagle3 draft cannot resume from a recurrent/hybrid checkpoint, forcing full re-processing\n");
-                                        do_reset = true;
-                                    }
-
                                     if (!do_reset) {
                                         // restore the context checkpoint
                                         it->load_tgt(ctx_tgt,       slot.id, LLAMA_STATE_SEQ_FLAGS_PARTIAL_ONLY);
                                         it->load_dft(ctx_dft.get(), slot.id, LLAMA_STATE_SEQ_FLAGS_PARTIAL_ONLY);
+                                        // restore the draft's deferred boundary (only eagle3 needs it; no-op otherwise)
+                                        common_speculative_set_deferred_boundary(spec.get(), slot.id, it->pos_max, it->data_dft_boundary_g_embd);
 
                                         pos_next = std::min(pos_next, std::max(it->pos_min + 1, it->pos_max));
                                         n_past   = std::min(slot.prompt.tokens.size_up_to_pos(pos_next), (size_t) it->n_tokens);

